interactive: Clean up natural_geometry hacks

jlindgren90 opened the pull request:
There are a couple of sketchy things going on here:

- interactive_begin() pokes its own values into view->natural_geometry to force view_maximize() to set a particular geometry.
- interactive_end() "fixes" view->natural_geometry after calling view_maximize() to save the original geometry from the start of the interactive move/resize.

Let's adjust/expand the API of view.c so that the interactive.c can avoid this "back door" of overwriting view->natural_geometry directly.

This also addresses an issue where the natural_geometry was not saved correctly when dragging and snapping a window to a screen edge.

Fixes: #633

Consolatis commented on src/interactive.c:
```diff
@@ -112,26 +106,22 @@ interactive_end(struct view *view)
 
+			/*
+			 * Don't store natural geometry here (it was
+			 * stored already in interative_begin())
```
s/interative/interactive/

jlindgren90 replied:
Oops! Fixed.

Consolatis commented on src/interactive.c:
```diff
@@ -112,26 +106,22 @@ interactive_end(struct view *view)
+				view_snap_to_edge(view, "left", false);
 			} else if (cursor_x >= area->x + area->width - snap_range) {
-				view_snap_to_edge(view, "right");
+				view_snap_to_edge(view, "right", false);
```
Inline comment about the new `false` would nice in this function as well, same for `view_maximize()` further down.

jlindgren90 replied:
Added inline comments.

Consolatis commented on src/interactive.c:
```diff
@@ -54,29 +54,23 @@ interactive_begin(struct view *view, enum input_mode mode, uint32_t edges)
 		}
+	} else {
+		/* Store natural geometry at start of move/resize */
+		view_store_natural_geometry(view);
```
I think we only have to store the natural geometry when `mode == LAB_INPUT_STATE_MOVE`. Maybe we could switch the `mode == MOVE` check and the `view->maximized || view->tiled check` here so the `mode == MOVE` check is the outer one and the else branch here is for the inner check (`!maximized && !tiled`)

jlindgren90 replied:
Good call. I reorganized `interactive_begin()` so that there is now one `switch(mode)` rather than the various `if` statements.

Consolatis commented on src/view.c:
```diff
@@ -423,44 +423,64 @@ view_apply_unmaximized_geometry(struct view *view)
 	if (maximize) {
-		interactive_end(view);
-		if (!view->tiled) {
+		interactive_end(view, /*allow_snap*/ false);
```
I don't remember why we have to call `interactive_end()` here in the first place, if you do would you mind adding a comment here?

jlindgren90 replied:
Added comments.

Consolatis commented on src/view.c:
```diff
@@ -534,7 +554,7 @@ view_set_fullscreen(struct view *view, bool fullscreen,
 	}
 	if (fullscreen) {
-		interactive_end(view);
+		interactive_end(view, /*allow_snap*/ false);
```
Same here

jlindgren90 replied:
Added comments.

Consolatis commented on src/view.c:
```diff
@@ -762,8 +783,8 @@ view_snap_to_edge(struct view *view, const char *direction)
 		/* Unmaximize + keep using existing natural_geometry */
-		view_maximize(view, false);
-	} else if (!view->tiled) {
+		view_maximize(view, false, false);
```
inline comment would be nice

jlindgren90 replied:
Added inline comment.

jlindgren90 commented:
v2:

- Save the natural geometry and the tiled state of the view in
  interactive_begin() when starting to move the view.  When done,
  interactive_end() will update the tiled state if appropriate but
  *not* overwrite the natural geometry.
- Refactor interactive_begin() a bit.
- Split interactive_end() into interactive_finish() and
  interactive_cancel(). Split out snap_to_edge().

johanmalm commented:
Works well. Another great contribution :smile: Thanks.
And thanks for review @Consolatis

Consolatis commented:
Nice changes. Thanks :)
